### PR TITLE
changefeedccl: SQL evaluation for CDC

### PR DESCRIPTION
**Pull request by miretskiy**
Introduce `cdceval` package -- a library for expression evaluation
for CDC.

Changefeed users for a long time requested ability to emit only a
subset of columns. They have also requested ability to filter
out unwanted events (for example, filter out deletions).

This library aims to accomplish those goals.  However, instead of
focusing on a narrow use cases, which would usually be addressed via
addition of new `WITH` option (as done in
https://github.com/cockroachdb/cockroach/pull/80499), this library
aims to provide support for general expression evaluation.

`cdceval` library provides the following functionality:
  * Ability to evaluate predicates (filters) so that events may be
    filtered.
  * Ability to evaluate projection expressions (`select *`, `select a,
    b,c`, or even `select a + b - c as math_column`)
  * Ability to evaluate virtual compute columns (currently not
    implemented in this PR).

`cdceval` library reuses existing parsing and evaluation libraries, but
adopts them for CDC use case.  CDC events are row level events, and as
such, CDC expressions only make sense in the context of a single
row/single table.  In addition, because CDC events are at least once
semantics, the emitted events must not depend on volatile state.
In summary, any expression is supported except:
  * Volatile functions -- not supported
  * Stable functions, such as `now()`, `current_timestamp()`, etc are
    modified so that they return stable values -- namely events MVCC
    timestamp.
  * Multi row functions (aggregates, windowing functions) are
    disallowed.

`cdceval` also defined few custom, CDC specific functions, such as:
  * `cdc_prev()`: Returns the previous row values as a JSONB object.
  * `cdc_is_delete()`: Returns true if the row was deleted.
  * Others -- see `functions.go`

The follow PRs will add a "front end" to this library to enable creation
and management of predicated changefeeds.

Release Notes: None

### CODE REVIEW

**Comment by jordanlewis** [comment]
Wow, this is really neat! Seems like it will open a whole new world of possibility.

It's probably early days still but have you thought about using the vectorized execution engine for scalar expression evaluation rather than expr.Eval? It's much more efficient with respect to garbage production and CPU usage, especially for cases like this where you'll be running a pre-decided set of projections and filters against many rows. If you already have "batches" to process it might be a very natural fit. Sorry for the drive by comment, but I couldn't resist.

**Comment by miretskiy** [comment]
> Wow, this is really neat! Seems like it will open a whole new world of possibility.
> 
Yes, I'm very excited about the potential here.

> It's probably early days still but have you thought about using the vectorized execution engine for scalar expression evaluation rather than expr.Eval? It's much more efficient with respect to garbage production and CPU usage, especially for cases like this where you'll be running a pre-decided set of projections and filters against many rows. If you already have "batches" to process it might be a very natural fit. Sorry for the drive by comment, but I couldn't resist.

It is early days, you're right.  Unfortunately, we are not processing batches -- just 1 row at a time, so I'm not sure vectorization would help.  However, one could imagine that at some point we might want to support "live query" type system where we could do joins against other tables as a result of an event.  In such cases, I would imagine batching & vectorization play important role.

Also, https://github.com/cockroachdb/cockroach/pull/80499 was the first foray for changefeeds into optimizer land.
I think I would to try adopting similar, narrow approach to projection evaluation -- basically, try to get some optimizer optimization goodies, just like we already get them for predicates.

**Review by HonoreDB — COMMENTED** [review]
> Reviewed 1 of 1 files at r1, 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81676)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @gh-casper, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 268 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81676#-N2qg-Em1Z7yB2cgQpJo:-N2qg-Em1Z7yB2cgQpJp:b-rzll9k) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd9f734ef86253e55e09a6fb49d602afb1ace0fa/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L268)):*
> ```Go
> 		return false, err
> 	}
> 	return d == tree.DBoolTrue, nil
> ```

Something to consider here is whether we want NULL to count as true or false in a predicate. In check constraints, NULL counts as true.
___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval.go` line 569 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81676#-N2qiXZ17z_wKeGQO9YT:-N2qiXZ17z_wKeGQO9YU:b-a77i7o) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd9f734ef86253e55e09a6fb49d602afb1ace0fa/pkg/ccl/changefeedccl/cdceval/expr_eval.go#L569)):*
> ```Go
> 	// Aggregates, generators and window functions are not supported.
> 	switch fnClass {
> 	case tree.AggregateClass, tree.GeneratorClass, tree.WindowClass:
> ```

It's occasionally been frustrating that you can't use a generator function inside an aggregator function in contexts that expect a scalar function--for example, `array_agg(json_array_elements(jsonCol))` to convert a json array into a SQL array. Do you know if it'd break anything here to allow that kind of expression? (Definitely not a blocker, regardless).
___
*[`pkg/ccl/changefeedccl/cdceval/functions_test.go` line 83 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/81676#-N2qmUgF1YL28bhy1Sp8:-N2qmUgF1YL28bhy1Sp9:b-6w82h3) ([raw file](https://github.com/cockroachdb/cockroach/blob/bd9f734ef86253e55e09a6fb49d602afb1ace0fa/pkg/ccl/changefeedccl/cdceval/functions_test.go#L83)):*
> ```Go
> 
> 		// When previous row is not set -- i.e. if running without diff, cdc_prev returns
> 		// null json.
> ```

Maybe we should instead walk the expression on startup looking for cdc_prev and set WithDiff to true if we find it, rather than making the user tell us they're interested in previous values in two places.

File: pkg/ccl/changefeedccl/cdceval/expr_eval.go [hunk 1]
```diff
@@ -0,0 +1,634 @@
...
+	}
+
+	// We must work harder.  Bind variables and resolve names.
+	expr, _ = tree.WalkExpr(&e.resolver, expr)
+	if e.resolver.err != nil {
+		return nil, e.resolver.err
+	}
+
+	if star, isStar := expr.(tree.UnqualifiedStar); isStar {
+		// Can't type check star -- we'll handle it later during eval.
```
Comment (HonoreDB) [inline pkg/ccl/changefeedccl/cdceval/expr_eval.go:390]: Could you use `types.AnyTuple` here?

File: pkg/ccl/changefeedccl/cdceval/functions.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	sessiondata.SearchPath
+}
+
+// Resolve implements tree.CustomFunctionDefinitionResolver
+func (cdcCustomFunctionResolver) Resolve(name string) *tree.FunctionDefinition {
+	fn, found := cdcFunctions[name]
+	if found {
+		return fn
+	}
+	fn, found = cdcFunctions[strings.ToLower(name)]
```
Comment (HonoreDB) [inline pkg/ccl/changefeedccl/cdceval/functions.go:166]: Would it make sense to do `cdcFunctions[name] = fn` if found so that we don't have to do this string call every time? Or is this resolution cached anyway?

**Comment by miretskiy** [comment]
> It's occasionally been frustrating that you can't use a generator function inside an aggregator function in contexts that expect a scalar function--for example, `array_agg(json_array_elements(jsonCol))` to convert a json array into a SQL array. Do you know if it'd break anything here to allow that kind of expression? (Definitely not a blocker, regardless).

Hmm...My thinking was that anything that operates in a context of a table or a set of rows should be disabled.
Does it have to be this way? Of course not.  We could for example implement `count(*)` if we wanted to (to return 1.  But I'm not sure that makes much sense.  I'm sure we could enable some uses... but, I just wonder if these make sense in the context of a single row event.

File: pkg/ccl/changefeedccl/cdceval/expr_eval.go (hunk 1 above)
Comment (miretskiy) [reply pkg/ccl/changefeedccl/cdceval/expr_eval.go:390]: Tbh, I'm not sure.  I see that this stuff is used when generating row(); i'm not sure how this would help here though.
* expansion still happens, and I happen to know exactly what types those expressions are...  Do you have any code gist where you think AnyTuple might come in handy?

Comment (HonoreDB) [reply pkg/ccl/changefeedccl/cdceval/expr_eval.go:390]: Thinking about it...probably no point, especially with no aggregate functions.

**Comment by HonoreDB** [comment]
>I just wonder if [aggregate functions] make sense in the context of a single row event.
I'm almost exclusively thinking about `unnest` and `json_array_elements`, which let you turn a single datum into a set you can act on with aggregate functions. Without access to set-returning and aggregate functions you tend to be out of luck if we haven't implemented the specific array function you're looking for. Like there's no way to delete all instances of a given json object from a json array in an update, or to create a check constraint that all elements of an array are unique. It's why I've been messing with our SQL builtins recently.

**Comment by miretskiy** [comment]
Added restriction to disallow subselects.  Just too big of a can of warms, and not clear if these are useful in a single row context. /cc @HonoreDB

**Review by HonoreDB — COMMENTED** [review]
> Reviewed 12 of 19 files at r3, 1 of 2 files at r7, 1 of 4 files at r8, 2 of 3 files at r9, 3 of 3 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81676)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @gh-casper, @HonoreDB, and @miretskiy)

___
*[`pkg/ccl/changefeedccl/cdceval/expr_eval_test.go` line 40 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/81676#-N35T1nd1Vcsf1QpT0S9:-N35T1nd1Vcsf1QpT0SA:b-ij78yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/89eea0a3ab0c43d37b3740844c3d4ad9f3ccc089/pkg/ccl/changefeedccl/cdceval/expr_eval_test.go#L40)):*
> ```Go
> 	"github.com/stretchr/testify/require"
> )
> ```

Few more fun tests: 
`SELECT IF(cdc_is_delete(),'deleted',a::string)` (I'm guessing this works)
`SELECT btrim(a)` (When does the unknown type signature error happen?)
`SELECT 1/(10-a)` (Do we fail gracefully when there's a runtime error evaluating SQL after some results are already returned?)
`SELECT parse_timetz('1:00-0')` (Ideally this succeeds since with a timezone specified it's not dependent on locale, but the implementation involves `GetTxnTimestamp` so I don't know if it will)

**Comment by miretskiy** [comment]
> Few more fun tests:
> `SELECT IF(cdc_is_delete(),'deleted',a::string)` (I'm guessing this works)

Added

> `SELECT btrim(a)` (When does the unknown type signature error happen?)
> `SELECT parse_timetz('1:00-0')` (Ideally this succeeds since with a timezone specified it's not dependent on locale, but the implementation involves `GetTxnTimestamp` so I don't know if it will)

Added both btrim + time parse test; as well as btrim(a).
Currently, error happens during projection because when you construct "evaluator", you don't have
a descriptor yet; I guess it's something that can be added later.

> `SELECT 1/(10-a)` (Do we fail gracefully when there's a runtime error evaluating SQL after some results are already returned?)

I have divide by 0 error tests.  There is really no such thing as some results returned.
It's an evaluation 1 row at a time.  What we need to have is a way to mark *any* evaluation error as a permanent error.
Why? Because (hopefully) we did our job and evaluation is "pure" -- only depends on row state -- thus, once you get
an error, you will always get an error.  So, must fail changefeed.  Poison pill handling will help with this problem.

**Review by HonoreDB — APPROVED** [review]
> Reviewed 1 of 3 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81676)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @gh-casper, @HonoreDB, and @miretskiy)

**Comment by miretskiy** [comment]
bors r=honoredb

**Comment by knz** [comment]
> Because (hopefully) we did our job and evaluation is "pure" -- only depends on row state

We have non-deterministic functions (e.g. `random`), so no

**Comment by miretskiy** [comment]
> 

We disallow those functions, @knz
